### ISSUE TITLE
Yolov8: How to get intermediate layers data?

### ISSUE THREAD

**Issue by lesept777** [ultralytics/ultralytics#9184]
### Search before asking

- [X] I have searched the YOLOv8 [issues](https://github.com/ultralytics/ultralytics/issues) and [discussions](https://github.com/ultralytics/ultralytics/discussions) and found no similar questions.

### Question

Hi
I'd like to know the input values of the last layer of my network, a softmax layer from the Yolov8.

The network is defined like this:

```
# Parameters
nc: 10 # number of classes
scales: # model compound scaling constants, i.e. 'model=yolov8n-cls.yaml' will call yolov8-cls.yaml with scale 'n'
  # [depth, width, max_channels]
  n: [0.30, 0.20, 1024]
  s: [0.33, 0.50, 1024]
  m: [0.67, 0.75, 1024]
  l: [1.00, 1.00, 1024]
  x: [1.00, 1.25, 1024]

# YOLOv8.0n backbone
backbone:
  # [from, repeats, module, args]
  - [-1, 1, Conv, [64, 3, 2]] # 0-P1/2
  - [-1, 1, Conv, [120, 3, 2]] # 1-P2/4
  - [-1, 3, C2f, [120, True]]
  - [-1, 1, Conv, [120, 3, 2]] # 3-P3/8
  - [-1, 6, C2f, [120, True]]
  - [-1, 1, Conv, [200, 3, 2]] # 5-P4/16
  - [-1, 6, C2f, [200, True]]

# YOLOv8.0n head
head:
  - [-1, 1, Classify, [nc]] # Classify

```
When I print the model, I don't see the softmax layer, but when I export it as onnx, I can see it...
![image](https://github.com/ultralytics/ultralytics/assets/30208898/6d9eac2a-519e-4443-b1e4-a121e4d34990)

How can I obtain the values before the softmax layer?

**Comment by glenn-jocher** [comment]
@lesept777 hello! 🙂 Getting the values before the softmax layer requires a slight modification to your model or utilizing hooks if you're working with PyTorch. Here's a quick example of how you can use a forward hook to get the output of the last layer before softmax:

```python
import torch

def get_features(name):
    def hook(model, input, output):
        print(f"{name} outputs:", output)
    return hook

# Assuming `model` is your YOLOv8 model
last_layer_name = 'the_name_of_the_layer_before_softmax'  # Fill this with the actual name
layer = getattr(model, last_layer_name)
handle = layer.register_forward_hook(get_features(last_layer_name))

# Run your model to see the outputs
# result = model(input)

handle.remove()  # Don't forget to remove the hook when you're done
```

You'll need to replace `'the_name_of_the_layer_before_softmax'` with the name or path to the layer you're interested in. This will print the output right before the softmax layer as it forwards.

For getting the layer output directly while training or inference, you might consider modifying your model definition to return this output as part of a tuple or another structure that fits your workflow. If you're working directly with ONNX models, examining layer outputs might require different tools specific to ONNX runtime or other frameworks you're using for inference. Hope this helps! 🚀

**Comment by Burhan-Q** [comment]
You could also try using `model.predict(image, embed=[N, M])` where `N` and `M` are the index values of the layers you want the intermediate output from
